### PR TITLE
NEW TakePOS compatibility with lots and serials

### PR DESCRIPTION
**Pull request by andreubisquerra**
Feature request https://github.com/Dolibarr/dolibarr/issues/21680

### CODE REVIEW

File: htdocs/takepos/invoice.php [hunk 1]
```diff
@@ -786,6 +861,11 @@ function fail($message)
 		$invoice->fetch($placeid);
 	}
 
+	if ($action=="setbatchid") {
+		$sql = "UPDATE ".MAIN_DB_PREFIX."facturedet set special_code=".((int) $batchid)." where rowid=".((int) $idoflineadded);
```
Comment (eldy) [inline htdocs/takepos/invoice.php:865]: The column special_code is already reserved to store information about the nature of line 1=transport, 2=ecotax, ...
So you will have problem if you reuse it to store batch number. So i added 2 new fields into llx_facturedet where you can store:
* field "batch" to store the batch/serial number (a varchar)
* field "fk_warehouse" to store the ID of the warehouse planed to use for stock decrementation

Then line 319, you can replace
`if ($line->special_code > 0) {
$prod_batch->fetch($line->special_code);`
with
`if ($line->batch && $line->fk_warehouse > 0) {
$prod_batch->find(0, '', '', $line->batch, $line->fk_warehouse);`
To get all information of batch stock to use for stock decrease.

**Comment by andreubisquerra** [comment]
special_code removed!
It's correct like this @eldy ?
I have checked it and it is working!

File: htdocs/takepos/invoice.php [hunk 2]
```diff
@@ -311,8 +311,25 @@ function fail($message)
...
 			dol_htmloutput_errors($invoice->error, $invoice->errors, 1);
 		}
+		// Update stock for batch products
+		if (isModEnabled('productbatch')) {
+			require_once DOL_DOCUMENT_ROOT . "/product/stock/class/mouvementstock.class.php";
+			$constantforkey = 'CASHDESK_ID_WAREHOUSE'.$_SESSION["takeposterminal"];
+			foreach ($invoice->lines as $line) {
+				if ($line->batch && $line->fk_warehouse > 0) {
+					$prod_batch = new Productbatch($db);
+					$prod_batch->fetch($line->batch);
```
Comment (eldy) [inline htdocs/takepos/invoice.php:321]: You must use here the method ->find to retreive the ProductBatch line from line->batch and line->fk_warehouse

**Comment by andreubisquerra** [comment]
Tested and works!, is now OK @eldy ?

**Pull request merged**